tangerine/installation.cpp

```diff
@@ -17,20 +17,39 @@
 
 #if !_WIN64
 #include <unistd.h>
+#include <sys/types.h>
+#include <pwd.h>
+#include <iostream>
+#include <cerrno>
+#include <cstring>
 #endif
 
+namespace fs = std::filesystem;
+
+#if !_WIN64
+const char *const tangerine_app_id =
+	// TODO: consider a reverse-dns name with the escaping recommended in:
+	// https://docs.gtk.org/gio/type_func.Application.id_is_valid.html
+	// See rationale in: https://docs.gtk.org/gtk4/migrating-3to4.html#set-a-proper-application-id
+	"tangerine";
+std::optional<fs::path> tangerine_get_xdg_state_home();
+#endif
 
 TangerinePaths::TangerinePaths(int argc, char* argv[])
 {
+	fs::path ExecutablePath;
 #if _WIN64
 	// TODO: This is fine for standalone builds, but we will want to do something
 	// else for future library builds.  Maybe GetModuleFileNameW?
-	ExecutablePath = std::filesystem::absolute(argv[0]);
+	ExecutablePath = fs::absolute(argv[0]);
 
 #elif EMBED_RACKET
-	ExecutablePath = std::filesystem::path(racket_get_self_exe_path(argv[0]));
+	// On Windows, `racket_get_self_exe_path()` returns UTF-8.
+	ExecutablePath = fs::path(racket_get_self_exe_path(argv[0]));
 
 #else
+	// The happy path is based on `racket_get_self_exe_path()`.
```

Comment:
What does "the happy path" mean here?

Reply:
I mean specifically that the code using `/proc/self/exe` is based on https://github.com/racket/racket/blob/2baec850fc6477a0c7b4c464981f0fd795b3990a/racket/src/start/self_exe.inc#L39-L83, but with the fallback to `std::filesystem::absolute(argv[0])` rather than `generic_get_self_path(exec_file)` (which tries several more things to get the best possible path).

This comment really should have been in the previous PR: I think I lost it while I was splitting the parts that worked out from the work in progress on Racket embedding.

Reply:
Ah ok.  I don't think the comment is necessary or accurate anymore as of 47d46eeafa02062424d5df89e0031b592154dcc0, since I rewrote that code only referencing API docs for `readlink` and the path name you were reading.

Reply:
Yes, I don't know that the basic idea of "look at `/proc/self/exe`" is legally significant, but IANAL.

Reply:
Never the less, I replaced just this code again so this is no longer an issue either way 24cfebffdc499982ae383aabbeb253d17d5e2c29